core/common/src/main/java/alluxio/resource/RoundRobinResourceQueue.java

```diff
@@ -0,0 +1,176 @@
+   * @param options construction options
+   */
+  public RoundRobinResourceQueue(DynamicResourcePool.Options options) {
+    mExecutor = Preconditions.checkNotNull(options.getGcExecutor());
+    mMaxCapacity = options.getMaxCapacity();
+    mGcFuture = mExecutor.scheduleAtFixedRate(
+        this::gcUnusedResources,
+        options.getInitialDelayMs(),
+        options.getGcIntervalMs(),
+        TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Acquire a {@link CloseableResource} from the queue.
+   * This method never fails or waits. If there is available slots in the queue,
+   * then a new resource is created and returned. Otherwise, a shared resource is
+   * returned.
+   * @return a closable resource
+   * @throws IOException when resource cannot be created
+   */
+  public CloseableResource<T> acquire() throws IOException {
+    try (LockResource r = new LockResource(mLock)) {
+      // we try to create new resource if there's available slots
+      // to avoid too much sharing that may hurt performance
+      if (mResourceQueue.size() < mMaxCapacity) {
+        T resource = createNewResource();
+        // Add new resource to the front of the deque so that it will
+        // be used first
+        mResourceQueue.addFirst(new CountingReference<>(resource, 0));
+      }
+
+      // here we are guaranteed that the queue is not empty, so the remove
+      // is safe
+      // always retrieve resource from the head of the queue and add it to the end
+      // so that resource will be dispatched in a round-robin manner
+      CountingReference<T> ref = mResourceQueue.removeFirst();
+      mResourceQueue.addLast(ref);
+      ref.reference();
+      return new CloseableResource<T>(ref.get()) {
+        @Override
+        public void closeResource() {
+          ref.dereference();
+        }
+      };
+    }
+  }
+
+  /**
+   * @return the number of resources the queue currently holds
+   */
+  public int size() {
+    try (LockResource r = new LockResource(mLock)) {
+      return mResourceQueue.size();
+    }
+  }
+
+  protected abstract T createNewResource() throws IOException;
+
+  protected abstract void closeResource(T resource) throws IOException;
+
+  @VisibleForTesting
+  void gcUnusedResources() {
+    ArrayList<T> resourceToGc = new ArrayList<>();
+    try (LockResource r = new LockResource(mLock)) {
+      if (mResourceQueue.size() < mMaxCapacity) {
+        return;
+      }
+      Iterator<CountingReference<T>> iterator = mResourceQueue.iterator();
+      while (iterator.hasNext()) {
+        CountingReference<T> nxt = iterator.next();
+        if (nxt.getRefCount() == 0) {
+          resourceToGc.add(nxt.get());
+          iterator.remove();
+        }
+      }
+    }
+
+    for (T resource: resourceToGc) {
+      try {
+        closeResource(resource);
+      } catch (IOException e) {
+        LOG.debug("Resource {} failed to be closed: {}", resource, e);
+      }
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
```

Comment:
Yeah this method is idempotent in itself. But you reminded me that the `CloseableResource` returned by the queue is not: if called twice, it would decrement the reference count twice. I think it makes sense to change `CloseableResource` to track a `mClosed` state.

Reply:
Besides that, `CloseableResource` should implement `AutoCloseable` instead. I can't really think of a valid use case for closing a `CloseableResouce` more than once. See this conversation https://github.com/Alluxio/alluxio/pull/15562#issuecomment-1139230119

Reply:
A use case where `Closeable` is more favorable than `AutoCloseable` is the use with guava's `Closer`, which requires `Closeable`s. We use `Closer`s quite a lot across the codebase.